universe/ValueRefs.h

```diff
@@ -760,6 +789,95 @@ void Variable<T>::serialize(Archive& ar, const unsigned int version)
         & BOOST_SERIALIZATION_NVP(m_return_immediate_value);
 }
 
+///////////////////////////////////////////////////////////
+// NamedRef                                              //
+///////////////////////////////////////////////////////////
+template <typename T>
+NamedRef<T>::NamedRef(const std::string& value_ref_name) :
+    m_value_ref_name(value_ref_name)
+{
+    DebugLogger() << "ctor(NamedRef<T>): " << typeid(*this).name() << " value_ref_name: " << m_value_ref_name;
+}
+
+template <typename T>
+bool NamedRef<T>::operator==(const ValueRef<T>& rhs) const
+{
+    if (&rhs == this)
+        return true;
+    if (typeid(rhs) != typeid(*this))
+        return false;
+    const NamedRef<T>& rhs_ = static_cast<const NamedRef<T>&>(rhs);
+    return (m_value_ref_name == rhs_.m_value_ref_name);
+}
+
+template <typename T>
+bool NamedRef<T>::RootCandidateInvariant() const
+{ return GetValueRef() ? GetValueRef()->RootCandidateInvariant() : true; }
```

Comment:
needs reworking into set in constructor after rebasing...

Reply:
thanks for the warning, i am fixing the code after rebasing right now

Reply:
ah sh... i cant really on checking the valueref at construction time-that valueref does not necessarily exist already.

Reply:
hmm... will need to make `*Invariant` functions virtual then...

Reply:
I think for the moment i do not have a use case for non-invariant value refs. So I will simply log a warning, when one registers a non-invariant value ref.